src/comps/qb.cc

```diff
@@ -0,0 +1,194 @@
+        this->curr_lim = std::min(4 * block_sz, k);
+        // No need for data movement in this case
+        RandLAPACK::comps::util::resize<T>(m * this->curr_lim, Q);
+        RandLAPACK::comps::util::resize<T>(this->curr_lim * n, B);
+    }
+    else
+    {
+        this->curr_lim = k;
+    }
+    
+    // Copy the initial data to avoid unwanted modification TODO #1
+    std::vector<T> A_cpy (m * n, 0.0);
+    T* A_cpy_dat = A_cpy.data();
+    lacpy(MatrixType::General, m, n, A_dat, m, A_cpy_dat, m);
+
+    T norm_B = 0.0;
+    T prev_err = 0.0;
+    T approx_err = 0.0;
+
+    T* Q_gram_dat;
+    T* Q_i_gram_dat;
+
+    if(this->orth_check)
+    {
+        Q_gram_dat = RandLAPACK::comps::util::resize(this->curr_lim * this->curr_lim, this->Q_gram);
+        Q_i_gram_dat = RandLAPACK::comps::util::resize(block_sz * block_sz, this->Q_i_gram);
+    }
+
+    T* QtQi_dat = RandLAPACK::comps::util::resize(this->curr_lim * block_sz, this->QtQi);
+    T* Q_i_dat = RandLAPACK::comps::util::resize(m * block_sz, this->Q_i);
+    T* B_i_dat = RandLAPACK::comps::util::resize(block_sz * n, this->B_i);
+
+    T* Q_dat = Q.data();
+    T* B_dat = B.data();
+    
+    while(k > curr_sz)
+    {
+        // Dynamically changing block size
+        block_sz = std::min(block_sz, k - curr_sz);
+        next_sz = curr_sz + block_sz;
+
+        // Make sure we have enough space for everything
+        if(next_sz > this->curr_lim)
+        {
+            this->curr_lim = std::min(2 * this->curr_lim, k);
+            Q_dat = RandLAPACK::comps::util::resize(this->curr_lim * m, Q);
+            B_dat = RandLAPACK::comps::util::row_resize(curr_sz, n, B, this->curr_lim);
+            QtQi_dat = RandLAPACK::comps::util::resize(this->curr_lim * block_sz, QtQi);
+            Q_gram_dat = RandLAPACK::comps::util::resize(this->curr_lim * this->curr_lim, Q_gram);
+        }
+
+        // Calling RangeFinder
+        this->RF_Obj.call(m, n, A_cpy, block_sz, this->Q_i);
+
+        if(this->orth_check)
+        {
+            if (RandLAPACK::comps::util::orthogonality_check(m, block_sz, block_sz, Q_i, Q_i_gram, this->verbosity))
+            {
+                // Lost orthonormality of Q
+                RandLAPACK::comps::util::row_resize(this->curr_lim, n, B, curr_sz);
+                k = curr_sz;
+                this->termination = 4;
+                return;
+            }
+        }
+
+        // No need to reorthogonalize on the 1st pass
+        if(curr_sz != 0)
+        {
+            // Q_i = orth(Q_i - Q(Q'Q_i))
+            gemm(Layout::ColMajor, Op::Trans, Op::NoTrans, curr_sz, block_sz, m, 1.0, Q_dat, m, Q_i_dat, m, 0.0, QtQi_dat, this->curr_lim);
+            gemm(Layout::ColMajor, Op::NoTrans, Op::NoTrans, m, block_sz, curr_sz, -1.0, Q_dat, m, QtQi_dat, this->curr_lim, 1.0, Q_i_dat, m);
+
+            // If CholQR failed, will use HQR
+            if(this->Orth_Obj.call(m, block_sz, this->Q_i))
+                this->Orth_Obj.call(m, block_sz, this->Q_i);
```

Comment:
This is making an assumption about the semantics of ``Orth_Obj.call(...)``. Specifically, it assumes that a stateful variable is set so that a second call to the same function can produce different results. My preference would be for ``Orth_Obj.call`` to be responsible for falling back on a backup algorithm if its primary algorithm failed.